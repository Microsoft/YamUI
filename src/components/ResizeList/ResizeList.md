### Examples

Basic usage:

```js { "props": { "data-description": "basic", "data-action-states": "[{\"action\":\"none\",\"wait\":\"50\"}]" } }
const { ButtonColor } = require('../Button');
const { BorderType } = require('../Image');

const items = [
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
];
const actionItem = <Image source="logo.png" description="Some action" borderType={BorderType.ROUND} height={32} width={32} />;

<ResizeList items={items} actionItem={actionItem} />
```

With a custom gutter size:

```js { "props": { "data-description": "large gutter", "data-action-states": "[{\"action\":\"none\",\"wait\":\"50\"}]" } }
const { GutterSize } = require('.');
const { ButtonColor } = require('../Button');
const { BorderType } = require('../Image');

const items = [
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
  <Image source={user.imageUrl} description={user.name} borderType={BorderType.ROUND} height={32} width={32} />,
];
const actionItem = <Image source="logo.png" description="Some action" borderType={BorderType.ROUND} height={32} width={32} />;

<ResizeList items={items} actionItem={actionItem} gutterSize={GutterSize.LARGE} />
```
