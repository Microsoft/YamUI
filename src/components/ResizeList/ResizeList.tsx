/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import { ResizeGroup } from 'office-ui-fabric-react/lib/ResizeGroup';
import { GutterSize, LayoutList, LayoutListItem, LayoutListProps } from '../LayoutList';

export { GutterSize };

interface ResizeListData {
  count: number;
  cacheKey: string;
}

export interface ResizeListProps extends BaseComponentProps {
  /**
   * A list of child elements to render. Items which don't fit the container width will not be rendered.
   */
  items: React.ReactNode[];

  /**
   * An additional item which will always be rendered at the end of th elist.
   */
  actionItem?: React.ReactNode;

  /**
   * Amount of spacing between list items. Defaults to 4px.
   * @default GutterSize.XSMALL
   */
  gutterSize?: LayoutListProps['gutterSize'];
}

/**
 * Describe what `ResizeList` does. This will also appear as part of this component's
 * documentation.
 */
export default class ResizeList extends React.Component<ResizeListProps> {
  public render() {
    const data: ResizeListData = {
      count: this.props.items.length,
      cacheKey: this.props.items.length.toString(),
    };

    return (
      <ResizeGroup
        data={data}
        onRenderData={this.onRenderData}
        onReduceData={this.onReduceData}
        onGrowData={this.onGrowData}
      />
    );
  }

  private onRenderData = (data: ResizeListData) => {
    const renderItems = this.props.items.slice(0, data.count);
    const actionItem = this.props.actionItem ? <LayoutListItem>{this.props.actionItem}</LayoutListItem> : null;

    return (
      <LayoutList
        className={join(['y-resizeList', this.props.className])}
        direction="horizontal"
        gutterSize={this.props.gutterSize}
      >
        {renderItems.map((item, index) => <LayoutListItem key={index}>{item}</LayoutListItem>)}
        {actionItem}
      </LayoutList>
    );
  };

  private onGrowData = (data: ResizeListData): ResizeListData | undefined => {
    // Prevent infinite loop when all items are rendered
    if (data.count >= this.props.items.length) {
      return undefined;
    }

    const count = data.count + 1;
    return {
      count,
      cacheKey: count.toString(),
    };
  };

  private onReduceData = (data: ResizeListData): ResizeListData | undefined => {
    // Prevent infinite loop when too small
    if (data.count === 0) {
      return undefined;
    }

    const count = data.count - 1;
    return {
      count,
      cacheKey: count.toString(),
    };
  };
}
