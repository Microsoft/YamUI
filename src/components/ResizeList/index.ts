/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './ResizeList';
export * from './ResizeList';
