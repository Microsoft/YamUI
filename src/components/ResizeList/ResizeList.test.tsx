/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ResizeList, { ResizeListProps, GutterSize } from './index';
import { ResizeGroup } from 'office-ui-fabric-react/lib/ResizeGroup';

describe('<ResizeList />', () => {
  let component: ShallowWrapper<ResizeListProps>;
  let items: React.ReactNode[];
  let actionItem: React.ReactNode;

  beforeEach(() => {
    items = [<div key="1">1</div>, <div key="2">2</div>, <div key="3">3</div>, <div key="4">4</div>];
    actionItem = <div>action></div>;
  });

  describe('with items', () => {
    beforeEach(() => {
      component = shallow(<ResizeList items={items} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with action item', () => {
    beforeEach(() => {
      component = shallow(<ResizeList items={items} actionItem={actionItem} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<ResizeList items={items} className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with gutter size', () => {
    beforeEach(() => {
      component = shallow(<ResizeList items={items} gutterSize={GutterSize.XLARGE} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when resizing the component', () => {
    let data: { count: number; cacheKey: string } | undefined;

    describe('down to the next smaller size', () => {
      beforeEach(() => {
        data = undefined;
        component = shallow(<ResizeList items={items} />);
        const onReduceData = component.find(ResizeGroup).prop('onReduceData');
        data = onReduceData({ count: 4, cacheKey: '4' });
      });

      it('onReduceData returns the correct object', () => {
        expect(data).toEqual({ count: 3, cacheKey: '3' });
      });
    });

    describe('down to zero', () => {
      beforeEach(() => {
        data = undefined;
        component = shallow(<ResizeList items={items} />);
        const onReduceData = component.find(ResizeGroup).prop('onReduceData');
        data = onReduceData({ count: 0, cacheKey: '0' });
      });

      it('onReduceData returns undefined', () => {
        expect(data).toBeUndefined();
      });
    });

    describe('up to the next larger size', () => {
      beforeEach(() => {
        data = undefined;
        component = shallow(<ResizeList items={items} />);
        const onGrowData = component.find(ResizeGroup).prop('onGrowData') as any;
        data = onGrowData({ count: 3, cacheKey: '3' });
      });

      it('onGrowData returns the correct object', () => {
        expect(data).toEqual({ count: 4, cacheKey: '4' });
      });
    });

    describe('larger than the number of items rendered', () => {
      beforeEach(() => {
        data = undefined;
        component = shallow(<ResizeList items={items} />);
        const onGrowData = component.find(ResizeGroup).prop('onGrowData') as any;
        data = onGrowData({ count: 4, cacheKey: '4' });
      });

      it('onGrowData returns undefined', () => {
        expect(data).toBeUndefined();
      });
    });
  });

  describe('onRenderData prop', () => {
    beforeEach(() => {
      component = shallow(<ResizeList items={items} />);
    });

    it('matches its snapshot', () => {
      expect(component.find(ResizeGroup).prop('onRenderData')({ count: 4, cacheKey: '4' })).toMatchSnapshot();
    });

    describe('with actionItem', () => {
      beforeEach(() => {
        component = shallow(<ResizeList items={items} actionItem={actionItem} />);
      });

      it('matches its snapshot', () => {
        expect(component.find(ResizeGroup).prop('onRenderData')({ count: 4, cacheKey: '4' })).toMatchSnapshot();
      });
    });
  });
});
