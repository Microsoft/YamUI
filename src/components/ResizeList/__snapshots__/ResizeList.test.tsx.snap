// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResizeList /> onRenderData prop matches its snapshot 1`] = `
<LayoutList
  className="y-resizeList"
  direction="horizontal"
>
  <LayoutListItem>
    <div>
      1
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      2
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      3
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      4
    </div>
  </LayoutListItem>
</LayoutList>
`;

exports[`<ResizeList /> onRenderData prop with actionItem matches its snapshot 1`] = `
<LayoutList
  className="y-resizeList"
  direction="horizontal"
>
  <LayoutListItem>
    <div>
      1
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      2
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      3
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      4
    </div>
  </LayoutListItem>
  <LayoutListItem>
    <div>
      action&gt;
    </div>
  </LayoutListItem>
</LayoutList>
`;

exports[`<ResizeList /> with action item matches its snapshot 1`] = `
<StyledCustomizedResizeGroup
  data={
    Object {
      "cacheKey": "4",
      "count": 4,
    }
  }
  onGrowData={[Function]}
  onReduceData={[Function]}
  onRenderData={[Function]}
/>
`;

exports[`<ResizeList /> with additional className matches its snapshot 1`] = `
<StyledCustomizedResizeGroup
  data={
    Object {
      "cacheKey": "4",
      "count": 4,
    }
  }
  onGrowData={[Function]}
  onReduceData={[Function]}
  onRenderData={[Function]}
/>
`;

exports[`<ResizeList /> with gutter size matches its snapshot 1`] = `
<StyledCustomizedResizeGroup
  data={
    Object {
      "cacheKey": "4",
      "count": 4,
    }
  }
  onGrowData={[Function]}
  onReduceData={[Function]}
  onRenderData={[Function]}
/>
`;

exports[`<ResizeList /> with items matches its snapshot 1`] = `
<StyledCustomizedResizeGroup
  data={
    Object {
      "cacheKey": "4",
      "count": 4,
    }
  }
  onGrowData={[Function]}
  onReduceData={[Function]}
  onRenderData={[Function]}
/>
`;
